### DISCUSSION TITLE
Using Firedrake in google colab

### DISCUSSION THREAD

**Discussion by Arijit-Majumdar** [firedrakeproject/firedrake#3302] [Firedrake support] [answered]
Hi,
I want to use Firedrake with google colab. I found some colab notebooks on the web which installs firedrake during the beginning of the notebook using the following lines of code

`try:`
`    import google.colab`  
`except ImportError:`
`    import firedrake`
`else:`
`    try:`
`        import firedrake`
`    except ImportError:`
`        if arch != "complex":`
`            !wget "https://fem-on-colab.github.io/releases/firedrake-install-real.sh" -O "/tmp/firedrake-install.sh" && bash "/tmp/firedrake-install.sh"`       
`    else:`
`            !wget "https://fem-on-colab.github.io/releases/firedrake-install-complex.sh" -O "/tmp/firedrake-install.sh" && bash "/tmp/firedrake-install.sh"`
`        import firedrake`
        
This would require installing firedrake every time I have to create a new notebook. Is it possible to install firedrake permanently in google colab, so that for any new colab notebook I can just call the library using `import firedrake`?

Thanks
Arijit

**Comment by connorjward** [comment]
I believe that the install is cached so the notebooks do not necessarily do a fresh install each time. @francesco-ballarin knows much more than me though.

Also, the install line here takes only a couple of minutes to complete, unlike the 1hr+ needed for a native Firedrake installation. Is that still an issue for your use case?

**Reply by francesco-ballarin** [reply, marked as answer]
Google Colab policies prevent permanent installation. As soon as you terminate the session (because you manually terminate it through the `Runtime` menu, or simply close the browser tab and don't reopen it till a few hours later) any local changes are lost. That includes `FEM on Colab` installations, but (notably) also any output (e.g., to `pvd` file for paraview) you may have created.

I am teaching with `FEM on Colab`, and none of the students has ever complained about needing to wait a couple of minutes every time. I realize that it would be easier to install `firedrake` once and forall, but the policy is out of our control and there is nothing we can do about that.

You may want to refer to https://fem-on-colab.github.io/packages.html for a slightly simpler installation code: 
```
try:
    import firedrake
except ImportError:
    !wget "https://fem-on-colab.github.io/releases/firedrake-install-real.sh" -O "/tmp/firedrake-install.sh" && bash "/tmp/firedrake-install.sh"
    import firedrake
```
the one you copied is more complicated because is used in the internal testing. Note that the `try/except` is there to make sure that if you *already* installed `firedrake` in the session you are currently in but for whatever reason you click again on the installation cell, it won't be installed again.
